pages/slice/language-reference/slice-grammar.md

```diff
@@ -2,14 +2,587 @@
+Optionally, a [tag] may be applied to the parameter, directly before its identifier, making this a [tagged parameter][tag].
+Additionally, [local attributes][attribute] and a [doc-comment] may be applied to the field in its prelude.
+
+```ebnf {% showTitle=false %}
+Parameter
+    : Prelude Tag? Identifier ":" "stream"? TypeRef
+    ;
+```
+
+Parameters also support the [`stream`][streamed-parameters-guide] modifier keyword, which may be applied to the parameter's type.
+Streamed parameters are only supported in [`Slice2`][compilation-mode-guide] mode, and within a list of parameters,
+at most one may be streamed and it must be the last parameter in that list.
+
+For additional information on parameters, see the [parameter][parameter-guide] pages.
+
+### Enum types
+
+Enum definitions consist of the `enum` keyword, followed by an [identifier], optionally followed by an [underlying type][type], and then the enum's body.
+Underlying types are specified by a single colon, followed by a type.
+Underlying types must be integral and non-optional. Enums have a default underlying type of [varint32][primitive] if none is specified.
+Enum bodies consist of a list of [enumerators][enumerator] wrapped in a pair of braces. These enumerators may be optionally separated by commas.
+Additionally, [local attributes][attribute] and a [doc-comment] may be applied to the enum in its prelude.
+
+```ebnf {% showTitle=false %}
+Enum
+    : Prelude "unchecked"? "enum" Identifier (":" TypeRef)? "{" UndelimitedList<Enumerator> "}"
+    ;
+```
+
+Enums also support the `unchecked` modifier keyword.
+An [unchecked-enum][unchecked-enum-guide] can be empty. An enum without this modifier must have at least one enumerator.
+
+For additional information on enums, see the [enum][enum-guide] page.
+
+### Enumerators
+
+Enumerators can only be defined within [enums][enum].
+Enumerator definitions consist of an [identifier], optionally followed by an enumerator value.
+Enumerator values consist of an equals sign, followed by a (possibly signed) [integer].
+Additionally, [local attributes][attribute] and a [doc-comment] may be applied to the enumerator in its prelude.
+
+```ebnf {% showTitle=false %}
+Enumerator
+    : Prelude Identifier ("=" SignedInteger)?
+    ;
+```
+
+The enum that an enumerator is defined within determines the allowed values.
+An enumerator's value must be within the bounds of the enum's underling type.
+
+For additional information on enumerators, see the [enumerator][enum-guide] page.
+
+### Custom types
+
+Custom type definitions consist of the `custom` keyword, followed by an [identifier].
+Additionally, [local attributes][attribute] and a [doc-comment] may be applied to the custom type in its prelude.
+
+```ebnf {% showTitle=false %}
+CustomType
+    : Prelude "custom" Identifier
+    ;
+```
+
+For a custom type to be mappable, it must have a `[x::type]` attribute applied to it, where `x` is the mapping language's prefix.
```

Comment:
Consider rephrasing the expression 'to be mappable' for clarity. How about:

For a custom type to be usable with a particular mapping language, it should have an attribute in the format [x::type], where x represents the prefix of that language.

Reply:
I rewrote it like you said, with some minor changes:
```
For a custom type to be usable with a particular language mapping, it must have an attribute of the form `[x::type]`,
where `x` is the Slice prefix of that language.
```

Let me know if you think it needs any more changes.